lib/vector-core/src/stream/driver.rs

```diff
@@ -0,0 +1,149 @@
+use std::{collections::HashMap, fmt, marker::PhantomData};
+
+use buffers::{Ackable, Acker};
+use futures::{stream::FuturesUnordered, FutureExt, Stream, StreamExt, TryFutureExt};
+use tokio::{pin, select, sync::oneshot};
+use tower::{Service, ServiceExt};
+use tracing::Instrument;
+
+use crate::event::{EventStatus, Finalizable};
+
+/// Drives the interaction between a stream of items and a service which processes them
+/// asynchronously.
+///
+/// `Driver`, as a high-level, facilitates taking items from an arbitrary `Stream` and pushing them
+/// through a `Service`, spawning each call to the service so that work can be run concurrently,
+/// managing waiting for the service to be ready before processing more items, and so on.
+///
+/// Additionally, `Driver` handles two event-specific facilities: finalization and acknowledgement.
+///
+/// This capability is parameterized so any implementation which can define how to interpret the
+/// response for each request, as well as define how many events a request is compromised of, can be
+/// used with `Driver`.
+pub struct Driver<St, Svc, Req>
+where
+    Svc: Service<Req>,
+{
+    input: St,
+    service: Svc,
+    acker: Acker,
+    _req: PhantomData<Req>,
+}
+
+impl<St, Svc, Req> Driver<St, Svc, Req>
+where
+    Svc: Service<Req>,
+{
+    pub fn new(input: St, service: Svc, acker: Acker) -> Self {
+        Self {
+            input,
+            service,
+            acker,
+            _req: PhantomData,
+        }
+    }
+}
+
+impl<St, Svc, Req> Driver<St, Svc, Req>
+where
+    St: Stream<Item = Req>,
+    Svc: Service<Req>,
+    Svc::Error: fmt::Debug + 'static,
+    Svc::Future: Send + 'static,
+    Svc::Response: AsRef<EventStatus>,
+    Req: Ackable + Finalizable,
+{
+    /// Runs the driver until the input stream is exhausted.
+    ///
+    /// All in-flight calls to the provided `service` will also be completed before `run` returns.
+    ///
+    /// # Errors
+    ///
+    /// No errors are currently returned.  Te return type is purely to simplify caller code, but may
+    /// return an error for a legitimate reason in the future.
+    pub async fn run(self) -> Result<(), ()> {
+        let in_flight = FuturesUnordered::new();
+        let mut pending_acks = HashMap::new();
```

Comment:
Small note that we could probably parameterize this with twox hash or some such if we wanted.

Reply:
I had the _exact_ same thought.  There's actually a [`nohash-hasher`](https://docs.rs/nohash-hasher) crate since our keys are already `u64`.  I guess if we've both had this thought, maybe I should just go ahead and do it.

Reply:
TIL about nohash-hasher.

Reply:
This is _almost_ a glorified `VecDeque`. I wonder if it could be emulated with a wrapper around `VecDequeue<Option<(u64, u64)>>` where inserting inserts "blanks" if it's not actually sequentially next.

Reply:
That's a good point regarding `VecDeque`.  I'll have to think about your idea... off the top of my head, I can't visualize it, so might just be one of those things I need to write down first to grok.

Reply:
In 05bc5f31a, I reworked this as a `BinaryHeap` because I didn't like the idea of having to insert blanks that we then had to do a linear scan to find when the call finishes and we want to mark the pending acknowledgement as completed.

Admittedly, though, the code using `BinaryHeap` is longer than the `HashMap` approach.